nmigen_boards/hadbadge.py

```diff
@@ -0,0 +1,168 @@
+            ),
+            Subsignal("rd",    Pins("P2")),
+            Subsignal("wr",    Pins("P4")),
+            Subsignal("rs",    Pins("P1")),
+            Subsignal("cs",    Pins("P3")),
+            Subsignal("id",    Pins("J4")),
+            Subsignal("rst",   Pins("H2")),
+            Subsignal("fmark", Pins("G1")),
+            Subsignal("blen",  Pins("P5")),
+            Attrs(IO_TYPE="LVCMOS33")
+        ),
+
+        Resource("spi_flash", 0, # clock needs to be accessed through USRMCLK
+            Subsignal("cs",   PinsN("R2")),
+            Subsignal("copi", Pins("W2")),
+            Subsignal("cipo", Pins("V2")),
+            Subsignal("wp",   Pins("Y2")),
+            Subsignal("hold", Pins("W1")),
+            Attrs(IO_TYPE="LVCMOS33")
+        ),
+        Resource("spi_flash_4x", 0, # clock needs to be accessed through USRMCLK
+            Subsignal("cs",   PinsN("R2")),
+            Subsignal("dq",   Pins("W2 V2 Y2 W1")),
+            Attrs(IO_TYPE="LVCMOS33")
+        ),
+
+        # SPI-connected PSRAM.
+        Resource("spi_ram_4x", 0,
+            Subsignal("cs",  PinsN("D20")),
+            Subsignal("clk", Pins("E20")),
+            Subsignal("dq",  Pins("E19 D19 C20 F19"), Attrs(PULLMODE="UP")),
+            Attrs(IO_TYPE="LVCMOS33", SLEWRATE="SLOW")
+        ),
+        Resource("spi_ram_4x", 1,
+            Subsignal("cs",   PinsN("F20")),
+            Subsignal("clk",  Pins("J19")),
+            Subsignal("dq",   Pins("J20 G19 G20 H20"), Attrs(PULLMODE="UP")),
+            Attrs(IO_TYPE="LVCMOS33", SLEWRATE="SLOW")
+        ),
+
+        SDRAMResource(0, clk="D11", cke="C11", cs="C7", we="B6", ras="D6", cas="A6",
+            ba="A7 C8", a="A8 D9 C9 B9 C14 E17 A12 B12 H17 G18 B8 A11 B11",
+            dq="C5 B5 A5 C6 B10 C10 D10 A9", dqm="A10",
+            attrs=Attrs(IO_TYPE="LVCMOS33", SLEWRATE="FAST")
+        )
+    ]
+
+    connectors = [
+        Connector("pmod", 0, "A15 C16 A14 D16 B15 C15 A13 B13"),
+        Connector("cartridge", 0,
+            "- - - - - - - - C5 B5 A5 C6 B6 A6 D6 C7 A7 C8 B8 A8 D9 C9 B9 A9" # continued:
+            " D10 C10 B10 A10 D11 C11 B11 A11 G18 H17 B12 A12 E17 C14"
+        ),
+
+        # SAO connectors names are compliant with the, erm, SAO 1.69 X-TREME standard.
+        # See: https://hackaday.com/2019/03/20/introducing-the-shitty-add-on-v1-69bis-standard/
+        Connector("sao", 0, {
+            "sda":   "B3", "scl":   "B2", "gpio0": "A2",
+            "gpio1": "A3", "gpio2": "B4", "gpio3": "A4"
+        }),
+        Connector("sao", 1, {
+            "sda":   "A16", "scl":   "B17", "gpio0": "B18",
+            "gpio1": "A17", "gpio2": "B16", "gpio3": "C17"
+        })
+    ]
+
+    def toolchain_prepare(self, fragment, name, **kwargs):
+        overrides = {
+            'ecppack_opts': '--compress --freq 38.8'
+        }
+        return super().toolchain_prepare(fragment, name, **overrides, **kwargs)
```

Comment:
This doesn't work if you try to override `ecppack_opts` externally. (That's equivalent to calling `toolchain_prepare(..., ecppack_opts="...", ecppack_opts="..."`, which is not valid in Python.)

Take a look at other boards with `toolchain_prepare` overrides.

Reply:
Oops; good catch. Copy-pasted the wrong block in there. :)